File: vayunx/_otel.py
"""OpenTelemetry providers: traces (spans) and process/machine gauges, exported over OTLP/HTTP."""

from __future__ import annotations

import logging
import platform
import socket
import threading
import time

from opentelemetry.exporter.otlp.proto.http.metric_exporter import OTLPMetricExporter
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.metrics import CallbackOptions, Observation
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import SpanProcessor, TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

from vayunx._cpuclock import thread_cpu_ns

CPU_ATTR = "vayunx.cpu_time_ms"


class _OncePerMinute(logging.Filter):
    """The OTLP exporters log an error on every failed export - once a second while the service is down.
    Offline mode must be quiet (spec A): let the first message through, then at most one a minute."""

    def __init__(self):
        super().__init__()
        self.next_at = 0.0
        self.suppressed = 0

    def filter(self, record: logging.LogRecord) -> bool:
        now = time.monotonic()
        if record.levelno >= logging.WARNING and now < self.next_at:
            self.suppressed += 1
            return False
        if record.levelno >= logging.WARNING:
            self.next_at = now + 60.0
            if self.suppressed:
                record.msg = f"{record.msg} ({self.suppressed} similar messages suppressed by vayunx)"
                self.suppressed = 0
        return True


EXPORT_LOG_FILTER = _OncePerMinute()
for _name in ("opentelemetry.exporter.otlp.proto.http.trace_exporter",
              "opentelemetry.exporter.otlp.proto.http.metric_exporter"):
    logging.getLogger(_name).addFilter(EXPORT_LOG_FILTER)


def resource_attrs(cfg, version: str) -> dict:
    attrs = {
        "service.name": cfg.service, "vayunx.run_id": cfg.run_id, "vayunx.variant": cfg.variant, "vayunx.phase": cfg.phase,
        "vayunx.sdk": f"vayunx-python/{version}", "telemetry.sdk.language": "python",
        "process.runtime.name": platform.python_implementation(), "process.runtime.version": platform.python_version(),
        "host.name": socket.gethostname(), "os.type": platform.system().lower(),
    }
    return {k: v for k, v in attrs.items() if v is not None}


class CpuTimeSpanProcessor(SpanProcessor):
    """Adds `vayunx.cpu_time_ms` (thread CPU time) to spans that start and end on the same thread.
    Spans that already carry the attribute (the SDK's retroactive crypto spans) are left alone."""

    def on_start(self, span, parent_context=None) -> None:
        try:
            if span.attributes and CPU_ATTR in span.attributes:
                return
            tid, c0 = threading.get_ident(), thread_cpu_ns()
            original_end = span.end

            def end(end_time=None):
                try:
                    if threading.get_ident() == tid:
                        span.set_attribute(CPU_ATTR, (thread_cpu_ns() - c0) / 1e6)
                except Exception:
                    pass
                original_end(end_time)

            span.end = end
        except Exception:
            pass

    def on_end(self, span) -> None:
        pass

    def shutdown(self) -> None:
        pass

    def force_flush(self, timeout_millis: int = 30000) -> bool:
        return True


def tracer_provider(cfg, attrs: dict) -> TracerProvider:
    tp = TracerProvider(resource=Resource.create(attrs), shutdown_on_exit=False)
    tp.add_span_processor(CpuTimeSpanProcessor())
    exporter = OTLPSpanExporter(endpoint=f"{cfg.endpoint}/v1/traces", timeout=cfg.flush_timeout_s, headers=cfg.auth_headers())
    tp.add_span_processor(BatchSpanProcessor(exporter, max_queue_size=2048, max_export_batch_size=512,
                                             schedule_delay_millis=1000, export_timeout_millis=int(cfg.flush_timeout_s * 1000)))
    return tp


class _GaugeReader:
    """Reads psutil once per collection and serves all gauges from that snapshot."""

    def __init__(self, activity):
        from vayunx_profiler_sdk.metrics import MachineReader, ProcessReader
        self.activity = activity
        self.proc, self.machine = ProcessReader(), MachineReader()
        self.proc.prime()
        self.machine.prime()
        self._snap: dict = {}
        self._at = 0.0
        self._lock = threading.Lock()

    def snapshot(self) -> dict:
        with self._lock:
            now = time.monotonic()
            if now - self._at > 0.25:
                p, m = self.proc.read(full=True), self.machine.read(full=False)
                self._snap = {**p, **m, "machine_other_cores_busy": self.machine.other_cores_busy(m, p)}
                self._at = now
            return self._snap

    def gauge(self, key: str):
        def callback(options: CallbackOptions):
            try:
                value = self.snapshot().get(key)
                if value is None:
                    return []
                category = "cryptographic" if self.activity.active() else "general"
                return [Observation(float(value), {"vayunx.category": category})]
            except Exception:
                return []
        return callback


GAUGES = {  # OTLP name -> key in the snapshot (see profiler_service/otlp.py GAUGES for the receiving side)
    "vayunx.process.cores_busy": "proc_cores_busy", "vayunx.process.cpu_time": "proc_cpu_time_s",
    "vayunx.process.rss_mib": "proc_rss_mib", "vayunx.process.peak_rss_mib": "proc_peak_rss_mib",
    "vayunx.process.threads": "proc_threads", "vayunx.machine.cpu_pct": "machine_cpu_pct",
    "vayunx.machine.other_cores_busy": "machine_other_cores_busy", "vayunx.machine.mem_available_mib": "machine_mem_available_mib",
}


def meter_provider(cfg, attrs: dict, activity) -> MeterProvider:
    exporter = OTLPMetricExporter(endpoint=f"{cfg.endpoint}/v1/metrics", timeout=cfg.flush_timeout_s, headers=cfg.auth_headers())
    reader = PeriodicExportingMetricReader(exporter, export_interval_millis=int(cfg.gauge_interval_s * 1000),
                                           export_timeout_millis=int(cfg.flush_timeout_s * 1000))
    mp = MeterProvider(resource=Resource.create(attrs), metric_readers=[reader], shutdown_on_exit=False)
    reader_obj = _GaugeReader(activity)
    meter = mp.get_meter("vayunx-python")
    for name, key in GAUGES.items():
        meter.create_observable_gauge(name, callbacks=[reader_obj.gauge(key)])
    return mp

File: sdk-node/src/otel.ts
/** OpenTelemetry providers (traces + process/runtime gauges) exported over OTLP/HTTP protobuf. */

import * as os from "node:os";
import { randomFill } from "node:crypto";
import { monitorEventLoopDelay, performance, type EventLoopUtilization, type IntervalHistogram } from "node:perf_hooks";

import { context, trace, type Tracer } from "@opentelemetry/api";
import { AsyncLocalStorageContextManager } from "@opentelemetry/context-async-hooks";
import { OTLPMetricExporter } from "@opentelemetry/exporter-metrics-otlp-proto";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-proto";
import { resourceFromAttributes } from "@opentelemetry/resources";
import { MeterProvider, PeriodicExportingMetricReader } from "@opentelemetry/sdk-metrics";
import { BasicTracerProvider, BatchSpanProcessor } from "@opentelemetry/sdk-trace-base";

import { authHeaders, type Config } from "./config";
import type { Attrs } from "./ship";

export function resourceAttrs(cfg: Config, version: string): Attrs {
  const attrs: Attrs = {
    "service.name": cfg.service, "vayunx.run_id": cfg.runId, "vayunx.variant": cfg.variant, "vayunx.phase": cfg.phase,
    "vayunx.sdk": `vayunx-node/${version}`, "telemetry.sdk.language": "nodejs", "process.runtime.name": "nodejs",
    "process.runtime.version": process.versions.node, "host.name": os.hostname(), "os.type": process.platform,
    "vayunx.runtime": process.env.NEXT_RUNTIME ? `next-${process.env.NEXT_RUNTIME}` : "node",
  };
  for (const k of Object.keys(attrs)) if (attrs[k] === undefined) delete attrs[k];
  return attrs;
}

export interface Tracing {
  provider: BasicTracerProvider;
  tracer: Tracer;
  registeredGlobal: boolean;
}

let contextManagerSet = false;

export function tracing(cfg: Config, attrs: Attrs, version: string): Tracing {
  // Offline mode stays quiet (spec A) because OTel diagnostics are a no-op unless the app set a diag logger;
  // we do not set one, so an app's own logger is never replaced.
  const exporter = new OTLPTraceExporter({ url: `${cfg.endpoint}/v1/traces`, timeoutMillis: cfg.flushTimeoutMs, headers: authHeaders(cfg) });
  const provider = new BasicTracerProvider({
    resource: resourceFromAttributes(attrs as Record<string, string | number | boolean>),
    spanProcessors: [new BatchSpanProcessor(exporter, {
      maxQueueSize: 2048, maxExportBatchSize: 512, scheduledDelayMillis: 1000, exportTimeoutMillis: cfg.flushTimeoutMs,
    })],
  });
  if (!contextManagerSet) {
    // true when no other context manager was registered first (then theirs keeps working for us too)
    contextManagerSet = context.setGlobalContextManager(new AsyncLocalStorageContextManager().enable()) || true;
  }
  const registeredGlobal = cfg.registerGlobal ? trace.setGlobalTracerProvider(provider) : false;
  return { provider, tracer: provider.getTracer("vayunx-node", version), registeredGlobal };
}

/** Reads process/machine/runtime state once per collection and serves every gauge from that snapshot. */
class GaugeReader {
  private lastCpu = process.cpuUsage();
  private lastWall = performance.now();
  private lastMachine = machineTimes();
  private elu: EventLoopUtilization = performance.eventLoopUtilization();
  private eld: IntervalHistogram = monitorEventLoopDelay({ resolution: 10 });
  private snap: Record<string, number> = {};
  private at = 0;
  private probeMs = 0;
  private probeTimer: NodeJS.Timeout;

  constructor(private active: () => boolean, intervalMs: number) {
    this.eld.enable();
    // Threadpool wait probe: a 1-byte randomFill runs on the libuv threadpool, so its latency is the
    // queue wait plus a negligible amount of work. Taken every interval; the gauge reports the latest.
    this.probeTimer = setInterval(() => {
      const t0 = performance.now();
      randomFill(Buffer.alloc(1), () => { this.probeMs = performance.now() - t0; });
    }, Math.max(250, intervalMs));
    this.probeTimer.unref();
  }

  close(): void {
    clearInterval(this.probeTimer);
    this.eld.disable();
  }

  snapshot(): Record<string, number> {
    const now = performance.now();
    if (now - this.at < 250) return this.snap;
    const cpu = process.cpuUsage();
    const wallUs = (now - this.lastWall) * 1000;
    const procCores = wallUs > 0 ? (cpu.user - this.lastCpu.user + cpu.system - this.lastCpu.system) / wallUs : 0;
    const m = machineTimes();
    const busy = m.busy - this.lastMachine.busy;
    const total = m.total - this.lastMachine.total;
    const machinePct = total > 0 ? (busy / total) * 100 : 0;
    const cores = os.cpus().length || 1;
    const elu = performance.eventLoopUtilization(this.elu);
    this.elu = performance.eventLoopUtilization();
    const p99 = this.eld.percentile(99) / 1e6;
    this.eld.reset();
    this.snap = {
      proc_cores_busy: procCores, proc_cpu_time_s: (cpu.user + cpu.system) / 1e6,
      proc_rss_mib: process.memoryUsage.rss() / 1048576, proc_peak_rss_mib: process.resourceUsage().maxRSS / 1024,
      machine_cpu_pct: machinePct, machine_other_cores_busy: Math.max(0, (machinePct / 100) * cores - procCores),
      machine_mem_available_mib: os.freemem() / 1048576, node_eventloop_delay_p99_ms: p99,
      node_eventloop_utilization: elu.utilization, node_threadpool_wait_ms: this.probeMs,
    };
    this.lastCpu = cpu;
    this.lastWall = now;
    this.lastMachine = m;
    this.at = now;
    return this.snap;
  }

  category(): string {
    return this.active() ? "cryptographic" : "general";
  }
}

function machineTimes(): { busy: number; total: number } {
  let busy = 0;
  let total = 0;
  for (const c of os.cpus()) {
    const t = c.times;
    const sum = t.user + t.nice + t.sys + t.idle + t.irq;
    total += sum;
    busy += sum - t.idle;
  }
  return { busy, total };
}

export const GAUGES: Record<string, [string, string]> = { // OTLP name -> [snapshot key, unit]; see profiler_service/otlp.py
  "vayunx.process.cores_busy": ["proc_cores_busy", "cores"],
  "vayunx.process.cpu_time": ["proc_cpu_time_s", "s"],
  "vayunx.process.rss_mib": ["proc_rss_mib", "MiB"],
  "vayunx.process.peak_rss_mib": ["proc_peak_rss_mib", "MiB"],
  "vayunx.machine.cpu_pct": ["machine_cpu_pct", "%"],
  "vayunx.machine.other_cores_busy": ["machine_other_cores_busy", "cores"],
  "vayunx.machine.mem_available_mib": ["machine_mem_available_mib", "MiB"],
  "vayunx.node.eventloop.delay_p99_ms": ["node_eventloop_delay_p99_ms", "ms"],
  "vayunx.node.eventloop.utilization": ["node_eventloop_utilization", "ratio"],
  "vayunx.node.threadpool.wait_ms": ["node_threadpool_wait_ms", "ms"],
};

export function gauges(cfg: Config, attrs: Attrs, active: () => boolean): { provider: MeterProvider; close(): void } {
  const reader = new GaugeReader(active, cfg.gaugeIntervalMs);
  const provider = new MeterProvider({
    resource: resourceFromAttributes(attrs as Record<string, string | number | boolean>),
    readers: [new PeriodicExportingMetricReader({
      exporter: new OTLPMetricExporter({ url: `${cfg.endpoint}/v1/metrics`, timeoutMillis: cfg.flushTimeoutMs, headers: authHeaders(cfg) }),
      exportIntervalMillis: cfg.gaugeIntervalMs, exportTimeoutMillis: Math.min(cfg.flushTimeoutMs, cfg.gaugeIntervalMs),
    })],
  });
  const meter = provider.getMeter("vayunx-node");
  for (const [name, [key, unit]] of Object.entries(GAUGES)) {
    meter.createObservableGauge(name, { unit }).addCallback((result) => {
      try {
        const v = reader.snapshot()[key];
        if (v !== undefined && Number.isFinite(v)) result.observe(v, { "vayunx.category": reader.category() });
      } catch { /* never throw into the host */ }
    });
  }
  return { provider, close: () => reader.close() };
}
